src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Formatting/Passes/New/CSharpFormattingPass.cs

```diff
@@ -0,0 +1,146 @@
+                    formattingChanges.Add(new TextChange(new TextSpan(originalStart, length), formattedCSharpText.ToString(TextSpan.FromBounds(formattedStart, formattedLine.End - lineInfo.FormattedOffsetFromEndOfLine))));
+
+                    if (lineInfo.CheckForNewLines)
+                    {
+                        Debug.Assert(lineInfo.FormattedLength == 0, "Can't have a FormattedLength if we're looking for new lines. The logic is incompatible.");
+                        Debug.Assert(lineInfo.FormattedOffsetFromEndOfLine == 0, "Can't have a FormattedOffsetFromEndOfLine if we're looking for new lines. The logic is incompatible.");
+
+                        // We assume Roslyn won't change anything but whitespace, so we can just apply the changes directly, but
+                        // it could very well be adding whitespace in the form of newlines, for example taking "if (true) {" and
+                        // making it run over two lines, or even "string Prop { get" and making it span three lines.
+                        // Since we assume Roslyn won't change anything non-whitespace, we just keep inserting the formatted lines
+                        // of C# until we match the original line contents.
+                        while (!FormattingUtilities.ContentEqualIgnoringWhitespace(changedText, originalStart, originalLine.End, formattedCSharpText, formattedStart, formattedLine.End))
+                        {
+                            // Sanity check: Because we're looking ahead through lines until the original line content is fully matches, we could loop forever if there is a bug somewhere
+                            Debug.Assert(
+                                FormattingUtilities.CountNonWhitespaceChars(changedText, originalStart, originalLine.End) >= FormattingUtilities.CountNonWhitespaceChars(formattedCSharpText, formattedStart, formattedLine.End),
+                                "Infinite loop in formatting! A bug in our visitor, or has Roslyn changed a non-whitespace char?");
+
+                            iFormatted++;
+                            formattedLine = formattedCSharpText.Lines[iFormatted];
+                            formattingChanges.Add(new TextChange(new(originalLine.EndIncludingLineBreak, 0), htmlIndentString + formattedCSharpText.ToString(formattedLine.SpanIncludingLineBreak)));
+                        }
+                    }
+                }
+            }
+
+            if (lineInfo.SkipNextLine)
+            {
+                iFormatted++;
+            }
+        }
+
+        // We're finished processing the original file, which means we've done all of the indentation for the file, and we've done
+        // the formatting changes for lines that are entirely C#, or start with C#, and lines that are Html or Razor. Now we process
+        // the "additional changes", which is formatting for C# that is inside Html, via implicit or explicit expressions.
```

Comment:
Does this include `@{ }` blocks?

Reply:
No, those are explicit _statements_. This is for handling C# that could appear at any point in a line, whereas whilst the `@{` itself could appear at any point, the `RazorFormattingPass` ensures the actual C# inside it will always be on a line by itself.